crates/admin/src/cluster_controller/scheduler.rs

```diff
@@ -174,6 +436,271 @@ impl<T: TransportConnect> Scheduler<T> {
+                            Err(PartitionConfigurationUpdate {
+                                current,
+                                next,
+                            })
+                        }
+                    }
+                }
+            }
+        }).await {
+            Ok(epoch_metadata) => {
+                info!("Successfully transitioned from partition processor configuration {} to {}", current_version, next_version);
+                let (_, _, current, next) = epoch_metadata.into_inner();
+                Ok(PartitionConfigurationUpdate {
+                    current,
+                    next,
+                })
+            }
+            Err(ReadModifyWriteError::FailedOperation(concurrent_update)) => {
+                Ok(concurrent_update)
+            }
+            Err(ReadModifyWriteError::ReadWrite(err)) => {
+                Err(err.into())
+            }
+        }
+    }
+
+    /// Checks whether the given partition requires reconfiguration. A partition requires
+    /// reconfiguration in the following cases:
+    ///
+    /// * next contains a replica that is no longer a valid worker candidate.
+    /// * current contains a replica that is no longer a valid worker candidate, and there is no
+    ///   ongoing reconfiguration happening already
+    ///
+    /// In this case, the method returns true, otherwise false.
+    fn requires_reconfiguration(
+        partition_state: &PartitionState,
+        nodes_config: &NodesConfiguration,
+    ) -> bool {
+        let current_requires_reconfiguration =
+            partition_state.current.replica_set().iter().any(|replica| {
+                !nodes_config
+                    .find_node_by_id(*replica)
+                    .map(|node_config| worker_candidate_filter(*replica, node_config))
+                    .unwrap_or_default()
+            });
+
+        let next_requires_reconfiguration = partition_state.next.as_ref().map(|next| {
+            next.replica_set().iter().any(|replica| {
+                !nodes_config
+                    .find_node_by_id(*replica)
+                    .map(|node_config| worker_candidate_filter(*replica, node_config))
+                    .unwrap_or_default()
+            })
+        });
+
+        let ongoing_reconfiguration = next_requires_reconfiguration.is_some();
+        let next_requires_reconfiguration = next_requires_reconfiguration.unwrap_or(false);
+
+        (current_requires_reconfiguration && !ongoing_reconfiguration)
+            || next_requires_reconfiguration
+    }
+
+    fn choose_partition_configuration(
+        partition_id: PartitionId,
+        nodes_config: &NodesConfiguration,
+        partition_replication: ReplicationProperty,
+        preferred_nodes: NodeSet,
+    ) -> Option<PartitionConfiguration> {
+        let options =
+            SelectorOptions::new(u64::from(partition_id)).with_preferred_nodes(preferred_nodes);
```

Comment:
As discussed previously, the downside of over-indexing on existing nodes are preferred will leave the cluster in a continuous imbalance given that we use partition_id as the hashing id.

Reply:
There is no one easy answer I'm afraid. But it could be generally better to _not_ use preferred nodes or have a heuristic of keeping 1 alive node from the previous set or something along those lines.

Reply:
Ok, then let's decouple this from this PR and do some experiments with different strategies to choose the best one.